cmd/login.go

```diff
@@ -102,208 +112,157 @@ type deviceTokenResponse struct {
-
 	accessTokenExpiry := ctxt.AccessTokenExpiry
 	if time.Now().After(accessTokenExpiry) {
 		if ctxt.RefreshToken == "" {
 			// We don't have a refresh token for this context, so we fail early
-			return "", fmt.Errorf("Could not login - invalid credentials. Please use the login command to refresh your credentials")
-		}
-
-		authProvider, err := getLoginInformation(http.DefaultClient, ctxt)
-
-		if err != nil {
-			cobra.CheckErr(fmt.Sprintf("Could not connect to %s - %s", ctxt.URL, err))
-			return "", err
+			cobra.CheckErr("Could not login - invalid credentials. Please use the login command to refresh your credentials")
 		}
 
 		// Access token has expired, we have to refresh it
+		authProvider := getLoginInformation(ctxt)
 		authProvider.grantType = "refresh_token"
 
 		if (authProvider.TokenURL != "") && (authProvider.ClientID != "") {
-
-			response, err := http.PostForm(authProvider.TokenURL,
-				url.Values{"grant_type": {authProvider.grantType},
-					"client_id":     {authProvider.ClientID},
-					"refresh_token": {ctxt.RefreshToken}})
-
-			if err != nil {
-				return "", fmt.Errorf("Cannot refresh access token - %s", err)
+			params := url.Values{
+				"refresh_token": {ctxt.RefreshToken},
 			}
-
-			var tokenResponse deviceTokenResponse
-			jsonDecoder := json.NewDecoder(response.Body)
-			if err := jsonDecoder.Decode(&tokenResponse); err != nil {
-				return "", fmt.Errorf("Cannot decode token response - %s", err)
+			tokenResponse := getTokenResponse(authProvider, params, ctxt, false)
+			if tokenResponse.ErrorString != "" {
+				logger.Warn("tokenResponse", log.String("error", tokenResponse.ErrorString))
+				cobra.CheckErr("oauth: Unexpected error from authentication provider")
 			}
 
-			switch tokenResponse.ErrorString {
-			case "authorization_pending":
-				// No op - we're waiting on the user to open the link and login
-			case "expired_token":
-				return "", fmt.Errorf("The login process was not completed in time - please login again")
-			case "access_denied":
-				return "", fmt.Errorf("Could not login - access was denied")
-			case "invalid_grant":
-				return "", fmt.Errorf("Could not login - expired credentials. Please use the login command to refresh your credentials")
-			case "":
-				// No Errors:
-				ctxt.AccessToken = tokenResponse.AccessToken
-				// Add a 10 second buffer to expiry to account for differences in clock time between client
-				// server and message transport time (oauth2 library does the same thing)
-				ctxt.AccessTokenExpiry = time.Now().Add(time.Second * time.Duration(tokenResponse.ExpiresIn-10))
-
-				// We also get an updated ID token, let's make sure we have the latest info
-				ParseIDToken(&tokenResponse, ctxt, authProvider.JwksURL)
-
-				fmt.Println(fmt.Sprintf("Successfully acquired new access token. Expiry: %s", ctxt.AccessTokenExpiry))
-
-				SetContext(ctxt, true)
-			}
+			ctxt.AccessToken = tokenResponse.AccessToken
+			ctxt.RefreshToken = tokenResponse.RefreshToken
```

Comment:
I haven't tested this, but are you sure that the token endpoint returns a (new?) refreshToken when we refresh the access token? I'm not sure this is in the standard and may just be an auth0 implementation detail. 
- [ ] Check if refresh token is updated/returned when we refresh the access token.

Reply:
I've looked further into this, and the specification says:

> If everything checks out, the service can [generate an access token](https://www.oauth.com/oauth2-servers/access-tokens/access-token-response/) and respond. The server may issue a new refresh token in the response, but if the response does not include a new refresh token, the client assumes the existing refresh token will still be valid.

So I think we should only update it, if it's been included in the response.

Reply:
Yes, and in fact it won't accept the old one anymore. Seems like a security feature to me.

Reply:
To be more specific, I think this should be:
```
if tokenResponse.RefreshToken != "" {
    ctxt.RefreshToken = tokenResponse.RefreshToken
}
```
So we don't replace the current valid refresh token with an empty one if it isn't included in the response.

Reply:
Done! 0d70853